README.md
# Melenium, planning stage

[![Status](https://img.shields.io/pypi/status/melenium)](https://pypi.org/project/melenium/)
[![Build Status](https://travis-ci.com/a-maliarov/melenium.svg?branch=main)](https://travis-ci.com/a-maliarov/melenium)
[![Code Coverage](https://img.shields.io/codecov/c/gh/a-maliarov/melenium?label=code%20coverage)](https://codecov.io/gh/a-maliarov/melenium)
[![Requirements Status](https://requires.io/github/a-maliarov/melenium/requirements.svg?branch=main)](https://requires.io/github/a-maliarov/melenium/requirements/?branch=main)
[![PyPi version](https://img.shields.io/pypi/v/melenium)](https://pypi.org/project/melenium/)


HISTORY.md
+ **Version 0.0.1**:
  1. Add `ChromeCapabilities` class.
+ **Version 0.0.2**:
  1. Add `ChromeDriver` class.
  2. Add `WaitFor` class.
  3. Move `WaitFor` and `Presets` into separated files.
+ **Version 0.1.0**:
  1. Add `ChromeSWB`, which now can extract XHR responses data.


tests/test_main.py
# -*- coding: utf-8 -*-

from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
from bs4 import BeautifulSoup as BS
from melenium import webdriver
import platform
import unittest
import sys
import os

#--------------------------------------------------------------------------------------------------------------

here = os.path.abspath(os.path.dirname(__file__))

caps = webdriver.ChromeCapabilities('maliarov')
caps.add_argument('--no-sandbox')
caps.add_argument('--headless')
driver = webdriver.Chrome(ChromeDriverManager("86.0.4240.22").install(), desired_capabilities=caps.desired)
driver.get('https://www.amazon.com/errors/validateCaptcha')

class TestMelenium(unittest.TestCase):

    def test_capabilities_add_extension(self):
        caps = webdriver.ChromeCapabilities()
        caps.add_extension(os.path.join(here, 'proxyautologin.crx'))
        self.assertIsNotNone(caps.desired['goog:chromeOptions']['extensions'])

    def test_capabilities_add_argument(self):
        caps = webdriver.ChromeCapabilities()
        caps.add_argument('test')
        self.assertIn('test', caps.desired['goog:chromeOptions']['args'])

    def test_capabilities_add_experimental_option(self):
        chrome_prefs = dict()
        chrome_prefs["profile.default_content_settings"] = {"images": 2}
        chrome_prefs["profile.managed_default_content_settings"] = {"images": 2}

        caps = webdriver.ChromeCapabilities()
        caps.add_experimental_option(chrome_prefs)
        self.assertEqual(chrome_prefs, caps.desired['goog:chromeOptions']['prefs'])

    def test_capabilities_set_user_agent(self):
        caps = webdriver.ChromeCapabilities()
        caps.set_user_agent("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36")
        self.assertIn("user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36", caps.desired['goog:chromeOptions']['args'])

    def test_capabilities_set_proxy(self):
        caps = webdriver.ChromeCapabilities()
        caps.set_proxy('test')
        self.assertEqual(caps.desired['proxy']['httpProxy'], 'test')

    def test_capabilities_set_download_folder(self):
        caps = webdriver.ChromeCapabilities()
        caps.set_download_folder('test')
        self.assertEqual(caps.desired['goog:chromeOptions']['prefs']['download.default_directory'], 'test')

    def test_capabilities_set_window_size(self):
        caps = webdriver.ChromeCapabilities()
        caps.set_window_size('1920x1080')
        self.assertIn('window-size=1920,1080', caps.desired['goog:chromeOptions']['args'])

    def test_capabilities_from_selenium_options(self):
        chrome_options = Options()
        chrome_options.add_argument('window-size=1920,1080')
        caps = webdriver.ChromeCapabilities.from_selenium_options(chrome_options)
        self.assertIn('window-size=1920,1080', caps.desired['goog:chromeOptions']['args'])

    #-------------------------------------------------------------------------

    def test_chromedriver_find(self):
        element = driver.find('img')
        self.assertIsNotNone(element)

    def test_chromedriver_find_2(self):
        element = driver.find('input', {'type': 'doesntexist'})
        self.assertIsNone(element)

    def test_chromedriver_find_element_by_bs(self):
        bs_element = BS(driver.page_source, features = 'html.parser').find('img')
        element = driver.find_element_by_bs(bs_element)
        self.assertIsNotNone(element)

    def test_chromedriver_find_element_by_bs_2(self):
        bs_element = BS(driver.page_source, features = 'html.parser').find('input', {'type': 'doesntexist'})
        element = driver.find_element_by_bs(bs_element)
        self.assertIsNone(element)

    def test_chromedriver_wait_for_phrase_in_link(self):
        driver.wait_for.phrase_in_link(10, 'validateCaptcha')
        self.assertIsNotNone(1)

    def test_chromedriver_wait_for_phrase_in_link_2(self):
        driver.wait_for.phrase_in_link(10, 'doesntexist')
        self.assertIsNotNone(1)

    def test_chromedriver_wait_for_element_in_dom(self):
        driver.wait_for.element_in_dom(10, 'div')
        self.assertIsNotNone(1)

    def test_chromedriver_wait_for_element_in_dom_2(self):
        driver.wait_for.element_in_dom(10, 'div', {'class': 'doesntexist'})
        self.assertIsNotNone(1)

#--------------------------------------------------------------------------------------------------------------

if __name__ == '__main__':
    unittest.main()

#--------------------------------------------------------------------------------------------------------------


melenium/wait_for.py
# -*- coding: utf-8 -*-

"""
melenium.wait_for
~~~~~~~~~~~~~~~~~

"""

__all__ = ['WaitFor']

import time

try:
    from bs4 import BeautifulSoup as BS
except ImportError:
    from beautifulsoup4 import BeautifulSoup as BS

#-----------------------------------------------------------------------------

class WaitFor(object):

    def __init__(self, driver):
        self._driver = driver

    def phrase_in_link(self, timeout, phrase='/'):
        counter = 0

        while phrase not in self._driver.current_url:
            time.sleep(1)
            counter += 1

            if counter == timeout:
                return None

    def element_in_dom(self, timeout, *argv, **kwargs):
        counter = 0

        while BS(self._driver.page_source, features = 'html.parser').find(*argv, **kwargs) == None:
            time.sleep(1)
            counter += 1

            if counter == timeout:
                return None

#-----------------------------------------------------------------------------


melenium/webdriver.py
# -*- coding: utf-8 -*-

"""
melenium.webdriver
~~~~~~~~~~~~~~~~~~

"""

__all__ = ['ChromeCapabilities', 'ChromeDriver', 'ChromeSWB']

from selenium.webdriver import Chrome as SeChrome

from seleniumwire.webdriver import Chrome as SwChrome
from seleniumwire.webdriver.request import InspectRequestsMixin
from seleniumwire.proxy.client import AdminClient

import pyaction as pa
import pickle

from .wait_for import WaitFor
from .common.capabilities import ChromeCapabilities

try:
    from bs4 import BeautifulSoup as BS
except ImportError:
    from beautifulsoup4 import BeautifulSoup as BS

#-----------------------------------------------------------------------------

class Chrome(SeChrome):

    def __init__(self, executable_path="chromedriver", port=0,
                 options=None, service_args=None,
                 desired_capabilities=None, service_log_path=None,
                 chrome_options=None, keep_alive=True):

        """
        Creates a new instance of the chrome driver.

        Starts the service and then creates new instance of chrome driver.

        :Args:
         - executable_path - path to the executable. If the default is used it assumes the executable is in the $PATH
         - port - port you would like the service to run, if left as 0, a free port will be found.
         - options - this takes an instance of ChromeOptions
         - service_args - List of args to pass to the driver service
         - desired_capabilities - Dictionary object with non-browser specific
           capabilities only, such as "proxy" or "loggingPref".
         - service_log_path - Where to log information from the driver.
         - chrome_options - Deprecated argument for options
         - keep_alive - Whether to configure ChromeRemoteConnection to use HTTP keep-alive.
        """

        super().__init__(executable_path, port, options, service_args, desired_capabilities, service_log_path, chrome_options, keep_alive)
        self._wait_for = WaitFor(self)

    def find(self, *argv, **kwargs):
        bs_element = BS(self.page_source, features = 'html.parser').find(*argv, **kwargs)

        if bs_element != None:
            xpath = pa.get_xpath(bs_element)
            return self.find_element_by_xpath(xpath)

        else:
            return None

    def find_element_by_bs(self, bs_element):
        if bs_element != None:
            xpath = pa.get_xpath(bs_element)
            return self.find_element_by_xpath(xpath)

        else:
            return None

    def upload_cookies(self, cookies_file, exclude=['expiry']):
        for cookie in pickle.load(open(cookies_file, "rb")):

            for key_ in exlude:
                if key_ in cookie:
                    del cookie[key_]

            self.add_cookie(cookie)

    @property
    def wait_for(self):
        return self._wait_for

#-----------------------------------------------------------------------------

class ChromeSWB(InspectRequestsMixin, Chrome):

    def __init__(self, *args, seleniumwire_options=None, **kwargs):

        if seleniumwire_options is None:
            seleniumwire_options = {}

        self._client = AdminClient()
        addr, port = self._client.create_proxy(
            port=seleniumwire_options.pop('port', 0),
            options=seleniumwire_options
        )

        if 'port' not in seleniumwire_options:
            try:
                capabilities = kwargs.pop('desired_capabilities')
            except KeyError:
                capabilities = ChromeCapabilities('empty').desired

            capabilities['proxy'] = {
                'proxyType': 'manual',
                'httpProxy': '{}:{}'.format(addr, port),
                'sslProxy': '{}:{}'.format(addr, port),
                'noProxy': ''
            }
            capabilities['acceptInsecureCerts'] = True

            kwargs['desired_capabilities'] = capabilities

        kwargs['desired_capabilities']['goog:chromeOptions']['args'].append('proxy-bypass-list=<-loopback>')

        super().__init__(*args, **kwargs)

    def quit(self):
        self._client.destroy_proxy()
        super().quit()

#-----------------------------------------------------------------------------


melenium/common/capabilities.py
# -*- coding: utf-8 -*-

"""
melenium.capabilities
~~~~~~~~~~~~~~~~~~~~~

Contains ChromeCapabilities.

"""

__all__ = ['ChromeCapabilities']

import base64

from .presets import PRESETS

#-----------------------------------------------------------------------------

class ChromeCapabilities(object):

    def __init__(self, preset='empty'):
        if isinstance(preset, str):
            self.desired = PRESETS[preset].copy()
        elif isinstance(preset, dict):
            self.desired = preset.copy()

    def add_argument(self, argument):
        self.desired['goog:chromeOptions']['args'].append(argument)

    def add_experimental_option(self, experimental_option):
        self.desired['goog:chromeOptions']['prefs'] = experimental_option

    def add_extension(self, extension):
        file_ = open(extension, 'rb')
        encoded_extension = base64.b64encode(file_.read()).decode('UTF-8')
        file_.close()

        self.desired['goog:chromeOptions']['extensions'].append(encoded_extension)

    def set_user_agent(self, user_agent):
        self.add_argument('user-agent={}'.format(user_agent))

    def set_proxy(self, proxy):
        proxy_types = ['httpProxy', 'ftpProxy', 'sslProxy']

        for type in proxy_types:
            self.desired['proxy'][type] = proxy

    def set_download_folder(self, folder_path):
        self.desired['goog:chromeOptions']['prefs']['download.default_directory'] = folder_path

    def set_window_size(self, window_size):
        self.add_argument('window-size={}'.format(window_size.replace("x", ",")))

    @classmethod
    def from_selenium_options(cls, selenium_options):
        current_options = selenium_options.to_capabilities()
        return cls(current_options)

#-----------------------------------------------------------------------------


melenium/common/presets.py
# -*- coding: utf-8 -*-

"""
melenium.presets
~~~~~~~~~~~~~~~~

List of chrome capabilities presets to be used by ChromeCapabilities.

"""

__all__ = ['PRESETS']

#-----------------------------------------------------------------------------

EMPTY_CAPABILITIES = {
    'browserName': 'chrome',
    'version': '',
    'platform': 'ANY',

    'goog:chromeOptions': {
        'prefs': dict(),
        'extensions': list(),
        'args': list()
    },

    'proxy': {
        'httpProxy': None,
        'ftpProxy': None,
        'sslProxy': None,
        'noProxy': None,
        'proxyType': 'MANUAL',
        'class': 'org.openqa.selenium.Proxy',
        'autodetect': False
    }
}

MALIAROV_BASIC_PRESET = {
    'browserName': 'chrome',
    'version': '',
    'platform': 'ANY',

    'goog:chromeOptions': {
        'prefs': dict(),
        'extensions': list(),
        'args': [
            'disable-auto-reload',
            'log-level=2',
            'disable-notifications',
            'start-maximized',
            'lang=en',
            'user-agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36"'
        ]
    },

    'proxy': {
        'httpProxy': None,
        'ftpProxy': None,
        'sslProxy': None,
        'noProxy': None,
        'proxyType': 'MANUAL',
        'class': 'org.openqa.selenium.Proxy',
        'autodetect': False
    }
}

PRESETS = {
    'empty': EMPTY_CAPABILITIES,
    'maliarov': MALIAROV_BASIC_PRESET
}

#-----------------------------------------------------------------------------
